### ISSUE TITLE
MSBuild runner does not automatically close `Xml` output file

### ISSUE THREAD

**Issue by dougbu** [xunit/xunit#1515]
1. Clone a repository that reads test result files after the `<xunit/>` target completes e.g. https://github.com/aspnet/AspNetWebStack
  This repo [includes](https://github.com/aspnet/AspNetWebStack/blob/da59ce7f2aa180515aee1e2650828a148899465d/tools/WebStack.xunit.targets#L11-L20) the following target:
  ``` xml
    <Target Name="Xunit">
        <!-- Override AppDomains default (set in xunit.runner.json). Need them with this runner. -->
        <xunit AppDomains="true" Assemblies="$(TestAssembly)" XmlV1="$(XmlPath)"/>
        <!-- Replace potentially illegal escaped characters (if they get through validations done during Save) -->
        <RegexReplace
            Files="$(XmlPath)"
            Find="&amp;#x(?&lt;char&gt;[0-9A-Fa-f]+);"
            Replace="\0x${char}"
            WarnOnNoMatch="false"/>
    </Target>
  ```
  `<RegexReplace/>` is an inline code target that [includes](https://github.com/aspnet/AspNetWebStack/blob/da59ce7f2aa180515aee1e2650828a148899465d/tools/WebStack.tasks.targets#L108) the following code:
  ``` c#
  string originalText = File.ReadAllText(fullPath);
  ```
2. Update the repo to use the 2.3.0 version of the xunit.runner.msbuild package.
2. If necessary, change the `<xunit/>` task to use `Xml` and not `XmlV1`.
3. Build a target that runs tests e.g. `UnitTest` in the [aspnet/AspNetWebStack](https://github.com/aspnet/AspNetWebStack) case

#### Expected
A clean build
#### Actual
A number of failures about file access:
```
...\AspNetWebStack\tools\WebStack.xunit.targets(15,9): error : The process cannot access the file
'...\AspNetWebStack\bin\CodeAnalysis\test\TestResults\WebMatrix.WebData.Test-XunitResults.xml'
because it is being used by another process.
```
#### Regression since 2.3 Beta 1
Haven't tested exhaustively but no similar problem is visible when using the `XmlV1` task parameter. The `TransformFactory` has the necessary `using` to close the result file immediately. Similarly, I found no similar problem when using the 2.0.0 MSBuild runner.

Examining the tags, it appears the problem began between 2.3 Beta 1 and Beta 2. See commit fd6f472a, [this line](https://github.com/xunit/xunit/commit/fd6f472adbaf74eb35db132cc58ff31161e2c4fa#diff-0901373d1b3e15f1ead1062915768a1bR207) in particular.
#### Likely root cause
Though `File.ReadAllText(...)` opens files with `FileShare.Read` and the MSBuild runner [uses](https://github.com/xunit/xunit/blob/8b89aa8bf7c7b4e67104897a807918f98c24299c/src/xunit.runner.msbuild/xunit.cs#L203) `new FileStream(Xml.GetMetadata("FullPath"), FileMode.OpenOrCreate)` which should use the same `FileShare.Read` flag, .NET disallows reading while the file is open for writing. The `<RegexReplace/>` task would need to open the file with `FileShare.ReadWrite`.

None of this should be necessary however. The real problem is that the runner does not reliably close the results file. That is, the `FileStream` opened [here](https://github.com/xunit/xunit/blob/8b89aa8bf7c7b4e67104897a807918f98c24299c/src/xunit.runner.msbuild/xunit.cs#L203) does not have a fixed lifetime. Need a `using`.

**Comment by dougbu** [comment]
Calling `GC.Collect()`, waiting a few seconds (I used five seconds), and retrying also works around this issue. This approach lets users stick with convenience methods like `File.ReadAllText(...)` but takes more lines and slows builds slightly.

**Comment by bradwilson** [comment]
I'm also mildly amused to see that the ASP.NET build is still using that not entirely great RegexReplace task that I wrote like a decade ago. :-p

**Comment by bradwilson** [comment]
Fixed by #1516.

**Comment by dougbu** [comment]
We try to keep the love alive 🥇
